File: sdk/python/grain/validate.py
"""JSON Schema validation for AgentSpec."""
from __future__ import annotations
import json
from dataclasses import dataclass
from pathlib import Path
from typing import Literal

import jsonschema

_SCHEMA_PATH = Path(__file__).parent / "schema.json"
_schema: dict | None = None


def _get_schema() -> dict:
    global _schema
    if _schema is None:
        _schema = json.loads(_SCHEMA_PATH.read_text())
    return _schema


@dataclass
class ValidationError:
    path: str
    message: str
    severity: Literal["error", "warning"] = "error"


def validate_spec(data: dict) -> list[ValidationError]:
    """Validate an AgentSpec dict against the JSON Schema.
    Returns a list of validation errors (empty = valid).
    """
    schema = _get_schema()
    validator = jsonschema.Draft202012Validator(schema)
    errors: list[ValidationError] = []
    for err in sorted(validator.iter_errors(data), key=lambda e: list(e.absolute_path)):
        path = "/" + "/".join(str(p) for p in err.absolute_path) if err.absolute_path else "/"
        errors.append(ValidationError(path=path, message=err.message))
    return errors


def assert_valid(data: dict) -> None:
    """Validate and raise if invalid."""
    errors = validate_spec(data)
    if errors:
        messages = "\n".join(f"  {e.path}: {e.message}" for e in errors)
        raise ValueError(f"AgentSpec validation failed:\n{messages}")

File: sdk/typescript/src/validate.ts
import Ajv2020 from "ajv/dist/2020.js"
import { readFileSync } from "node:fs"
import { fileURLToPath } from "node:url"
import { dirname, join } from "node:path"

const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)

export interface ValidationError {
  path: string
  message: string
  severity: "error" | "warning"
}

// Load schema — try package-local first (npm install), then monorepo path (development)
const schemaPath = (() => {
  const packageLocal = join(__dirname, "..", "agentspec.schema.json")
  const monorepo = join(__dirname, "..", "..", "..", "spec", "agentspec.schema.json")
  try {
    readFileSync(packageLocal, "utf-8")
    return packageLocal
  } catch {
    return monorepo
  }
})()
let _schema: object | null = null

function getSchema(): object {
  if (!_schema) {
    _schema = JSON.parse(readFileSync(schemaPath, "utf-8"))
  }
  return _schema!
}

let _validateFn: ReturnType<Ajv2020["compile"]> | null = null

function getValidator(): ReturnType<Ajv2020["compile"]> {
  if (!_validateFn) {
    const ajv = new Ajv2020({ allErrors: true, strict: false })
    _validateFn = ajv.compile(getSchema())
  }
  return _validateFn
}

/**
 * Validate an AgentSpec object against the JSON Schema.
 * Returns an array of validation errors (empty = valid).
 */
export function validateSpec(data: unknown): ValidationError[] {
  const validate = getValidator()
  const valid = validate(data)

  if (valid) return []

  return (validate.errors ?? []).map(err => ({
    path: err.instancePath || "/",
    message: err.message ?? "Unknown validation error",
    severity: "error" as const,
  }))
}

/**
 * Validate and throw if invalid.
 */
export function assertValid(data: unknown): void {
  const errors = validateSpec(data)
  if (errors.length > 0) {
    const messages = errors.map(e => `  ${e.path}: ${e.message}`).join("\n")
    throw new Error(`AgentSpec validation failed:\n${messages}`)
  }
}
